### DISCUSSION TITLE
How to use Context::include_bytes for Images ?

### DISCUSSION THREAD

**Discussion by Shrav108** [emilk/egui#5021] [Announcements]
I have an application where the GUI updates new plots on to the screen when a button is pressed. The plot gets saves a .jpeg internally and displays it on the app. I have the following code

```
ui.add(egui::Image::new("D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg")
                                    .max_height(340.0)
                                    .max_width(560.0)
                                    .maintain_aspect_ratio(true)
                                    .rounding(10.0));
```

In cargo.toml : 

```
egui_extras   = { version = "0.28.1", features = ["all_loaders"] }
```

When I run the app,  I get this message

![image](https://github.com/user-attachments/assets/5ab2dd87-c0fb-4904-bf2b-63010eb74815)

I also tried with : 
```
ui.add(egui::Image::from_uri("D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg"))
```
but still the same message.

How to include Context::include_byte or add more byte loaders ?

**Comment by emilk** [comment]
You need to use [`egui_extras::install_image_loaders`](https://docs.rs/egui_extras/latest/egui_extras/fn.install_image_loaders.html)

**Comment by Shrav108** [comment]
I have already put this function in my main code. I have a code below that shows the plot image when I press the button 
```
ui.add(egui::Image::new(egui::include_image!("D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg"))
                                .max_height(330.0)
                                .max_width(560.0)
                                .maintain_aspect_ratio(true)
                                .rounding(10.0));
```
The problem here is that : 
1. First time I press the button with some inputs, it shows the plotted image.
2. But when I press the button with different inputs
        -> _it saves the new plotted image internally(from plotly crate), but does not update on the app (the older plot is shown)._

The updated image has the same name i.e. overwritten.

Which is why I used : 

```
ui.add(egui::Image::new("D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg")
                                    .max_height(340.0)
                                    .max_width(560.0)
                                    .maintain_aspect_ratio(true)
                                    .rounding(10.0));
```
but when I use this code, I get : 
![362128313-5ab2dd87-c0fb-4904-bf2b-63010eb74815](https://github.com/user-attachments/assets/35d15466-81d2-4675-910f-4368d1c14a0b)

I think the egui::include_image! does not update new version of the same file name.

What should I do ?

**Reply by YgorSouza** [reply]
There are 4 problems:

- When using `include_image`, it embeds the image data into your executable at compile time. It won't load images at runtime. So you can't use that in your case
- To load images at runtime, you have to enable the `file` feature in egui_extras (although you enabled `all_loaders` so I guess that covers that)
- You also have to prepend `file://` to the file URI, like `format!("file://{path}")`
- The loader caches the image data using the URI as key, to avoid reloading it on every frame. So if the image changes, you have to tell egui to forget the previous image using [Context::forget_image](https://docs.rs/egui/0.28.1/egui/struct.Context.html#method.forget_image). I think this has to include the `file://` prefix as well, i.e., use the same URI you passed to `Image::new`

**Reply by Shrav108** [reply]
Thank you ! Your suggestion works 👍 

Here is the code : 

When I press the button, I put the following code : 
```
ctx.forget_image("file://D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg");
```

then display the new image with : 
```
ui.add(egui::Image::new("file://D:/XYZ/hydrostatic/Pics/gap_load_stiff_equal.jpeg")
                                .max_height(330.0)
                                .max_width(560.0)
                                .maintain_aspect_ratio(true)
                                .rounding(10.0));
```
